File: java/ex2.java
class Veiculo {
    private String marca;
    private String modelo;
    private int ano;
    private int velocidade; 

    public Veiculo(String marca, String modelo, int ano) {
        this.marca = marca;
        this.modelo = modelo;
        this.ano = ano;
        this.velocidade = 0; 
    }

    public void acelerar(int incremento) {
        this.velocidade += incremento;
        System.out.println("Acelerando. Velocidade atual: " + this.velocidade + " km/h");
    }

    public void frear(int decremento) {
        this.velocidade -= decremento;
        if (this.velocidade < 0) {
            this.velocidade = 0; 
        }
        System.out.println("Freando. Velocidade atual: " + this.velocidade + " km/h");
    }

    public void mostrarVelocidade() {
        System.out.println("Velocidade atual: " + this.velocidade + " km/h");
    }

    public void mostrarDetalhes() {
        System.out.println("Marca: " + this.marca + ", Modelo: " + this.modelo + ", Ano: " + this.ano);
    }

    public static void main(String[] args) {
        
        Veiculo veiculo1 = new Veiculo("Toyota", "Corolla", 2020);
        Veiculo veiculo2 = new Veiculo("Honda", "Civic", 2021);
        Veiculo veiculo3 = new Veiculo("Ford", "Focus", 2019);

        veiculo1.mostrarDetalhes();
        veiculo2.mostrarDetalhes();
        veiculo3.mostrarDetalhes();

        veiculo1.acelerar(50);
        veiculo1.frear(20);
        veiculo1.mostrarVelocidade();
        veiculo1.frear(40); 
    }
}

File: golang/ex2.go
package main

import (
	"fmt"
)

type Automovel struct {
	Marca      string
	Model      string
	Ano        int
	Velocidade int
}

func (a *Automovel) Acelerar(quantidade int) {
	a.Velocidade += quantidade
}

func (a *Automovel) DiminuirVelocidade(quantidade int) {
	if a.Velocidade-quantidade < 0 {
		a.Velocidade = 0
	} else {
		a.Velocidade -= quantidade
	}
}

func (a Automovel) MostrarVelocidade() {
	fmt.Printf("A velocidade atual do %s %s é: %d km/h\n", a.Marca, a.Model, a.Velocidade)
}

func main() {

	veiculo1 := Automovel{Marca: "Toyota", Model: "Corolla", Ano: 2020}
	veiculo2 := Automovel{Marca: "Ford", Model: "Focus", Ano: 2019}
	veiculo3 := Automovel{Marca: "Honda", Model: "Civic", Ano: 2021}

	veiculo1.Acelerar(50)
	veiculo1.MostrarVelocidade()

	veiculo2.Acelerar(30)
	veiculo2.MostrarVelocidade()

	veiculo3.Acelerar(70)
	veiculo3.MostrarVelocidade()

	veiculo1.DiminuirVelocidade(20)
	veiculo1.MostrarVelocidade()

	veiculo2.DiminuirVelocidade(50)
	veiculo2.MostrarVelocidade()

	veiculo3.DiminuirVelocidade(100)
	veiculo3.MostrarVelocidade()
}
